lightning/src/chain/chainmonitor.rs

```diff
@@ -163,43 +181,50 @@ impl<ChanSigner: ChannelKeys, C: Deref, T: Deref, F: Deref, L: Deref> ChainMonit
 				}
 			}
 		}
-		entry.insert(monitor);
-		Ok(())
-	}
-
-	/// Updates the monitor that watches the channel referred to by the given outpoint.
-	fn update_monitor(&self, outpoint: OutPoint, update: ChannelMonitorUpdate) -> Result<(), MonitorUpdateError> {
-		let mut monitors = self.monitors.lock().unwrap();
-		match monitors.get_mut(&outpoint) {
-			Some(orig_monitor) => {
-				log_trace!(self.logger, "Updating Channel Monitor for channel {}", log_funding_info!(orig_monitor));
-				orig_monitor.update_monitor(update, &self.broadcaster, &self.logger)
+		match self.persister.persist_new_channel(funding_outpoint, &monitor) {
+			Err(e) => {
+				log_error!(self.logger, "Failed to persist new channel data");
+				return Err(e);
 			},
-			None => Err(MonitorUpdateError("No such monitor registered"))
-		}
-	}
-}
-
-impl<ChanSigner: ChannelKeys, C: Deref + Sync + Send, T: Deref + Sync + Send, F: Deref + Sync + Send, L: Deref + Sync + Send> chain::Watch for ChainMonitor<ChanSigner, C, T, F, L>
-	where C::Target: chain::Filter,
-	      T::Target: BroadcasterInterface,
-	      F::Target: FeeEstimator,
-	      L::Target: Logger,
-{
-	type Keys = ChanSigner;
-
-	fn watch_channel(&self, funding_txo: OutPoint, monitor: ChannelMonitor<ChanSigner>) -> Result<(), ChannelMonitorUpdateErr> {
-		match self.add_monitor(funding_txo, monitor) {
-			Ok(_) => Ok(()),
-			Err(_) => Err(ChannelMonitorUpdateErr::PermanentFailure),
+			_ => {}
 		}
+		entry.insert(monitor);
+		Ok(())
 	}
 
 	fn update_channel(&self, funding_txo: OutPoint, update: ChannelMonitorUpdate) -> Result<(), ChannelMonitorUpdateErr> {
-		match self.update_monitor(funding_txo, update) {
-			Ok(_) => Ok(()),
-			Err(_) => Err(ChannelMonitorUpdateErr::PermanentFailure),
+		// Update the monitor that watches the channel referred to by the given outpoint.
+		let mut monitors = self.monitors.lock().unwrap();
+		if let Some(orig_monitor) = monitors.get_mut(&funding_txo) {
+			log_trace!(self.logger, "Updating Channel Monitor for channel {}", log_funding_info!(orig_monitor));
+			let mut should_persist = true;
+			let res = orig_monitor.update_monitor(&update, &self.broadcaster, &self.logger);
+			match res {
+				Err(MonitorUpdateError::PersistMonitor(msg)) => {
+					log_error!(self.logger, "{}", msg);
+				},
+				Err(MonitorUpdateError::NoPersistMonitor(msg)) => {
+					log_error!(self.logger, "{}", msg);
+					should_persist = false;
+				},
+				Ok(()) => {},
+			}
```

Comment:
See point here : https://github.com/rust-bitcoin/rust-lightning/pull/681#discussion_r498471668

I think NoPersistMonitor can disappear as a rejected state still _update_ ChannelMonitor (like in a case of receiving an update from N+1 when commitment N has already been broadcast) which is a different question that returning an error to off-chain coordinator.

Reply:
That does seem to be the case. @TheBlueMatt were you thinking that other non-persistence-requiring errors might be added in the future? Otherwise, it seems like we can get rid of the `Persist/NoPersist` thing [mentioned](https://github.com/rust-bitcoin/rust-lightning/pull/681#discussion_r485875141)

Reply:
I mean I guess we can get rid of it? I agree its awkward to persist the same monitor again after it failed to be updated, but doing something stupid but non-destructive in the error path isn't always the worst thing.

Reply:
Thinking it a bit, i slightly prefer to keep it because it makes the weird logic more self-documenting.

Reply:
> Thinking it a bit, i slightly prefer to keep it because it makes the weird logic more self-documenting.

It doesn't make sense to keep it and we should store if we take action on invalid state send us by the ChannelMonitor. A wrong provided revocation secret is a good example of such case where we might start to distrust the coordinator and close preemptively the channel.

The logic isn't weird we just have a distributed systems with ultimately two "masters", the off-chain coordinator and the blockchain itself, and you can't expect them to be synchronized.

It just a matter of how we explain validity. An update can be valid but still be received as obsolete due to block height clocks not being synchronized.

Reply:
Discussed on ldk-dev Slack, removed!